DOCS/tech-overview.txt

```diff
@@ -555,13 +555,13 @@ Condition variables
 -------------------
 
 They're used whenever a thread needs to wait for something, without nonsense
-like sleep calls or busy waiting. mpv uses the standard pthread API for this.
+like sleep calls or busy waiting. mpv uses the mp_thread API for this.
 There's a lot of literature on it. Read it.
```

Comment:
There isn't lots of literature on `mp_thread` API :)

Which brings up another point, we probably need some documentation on this new interface, i.e what type of guarantees it makes and doesn't make etc. We can go two routes:

1. Start fresh. Make the absolute minimum amount of guarantees that mpv code expects/needs and cut the churn.
2. Make it "based on pthreads" and document deviations.

The latter is probably easier to start off with, but pthreads have lots of subtle details that can be easy to overlook.

Reply:
As it is more or less pthreads document it based on that.
After this extra layer is here just because windows is missing it and somewhat difficult to implement in a good way directly.

Reply:
> The latter is probably easier to start off with, but pthreads have lots of subtle details that can be easy to overlook and/or get wrong.

The functionality that mpv uses is pretty basic, we just have mutex/condvar and thread, that's all.

>  After this extra layer is here just because windows is missing it and somewhat difficult to implement in a good way directly.

@DanOscarsson: Sorry, I don't understand what you mean. You mean that this extra layer is here only for windows, because it is hard to implement otherwise or you mean the new interface is hard to implement on windows directly. Both old and new version was implemented for win32.

And no, this is not a wrapper for Windows. The main point is to abstract implementation details of pthread and clock used. Use `mp_time` as an internal clock and convert only when needed. Of course there are some adjustment to make Windows wrapper lightweight without the need of tracking thread pids. The main difference that with pthread you can get current thread handle, but not pid and on Windows is the other way around, you can get current thread pid and not handle.

Reply:
> There isn't lots of literature on mp_thread API :)

Updated this line. I've also skimmed through the rest of the threading overview and it flows nicely. Our API is basic ans uses pretty standard primitives and nor the API, nor the users of it cares about small intricacies. I still don't feel like mpv's technical overview is the place to teach threading and in the same time while we define custom mp_thread now, all primitives of it are pretty standardized and knowledge transferable. I don't think we have to do full description of it and developers would be better of it they learn pthreads, std::thread or win32 threads first... and then usage of this mp_thread is quite standard.